design/thin-edge-actors-design.md

```diff
@@ -0,0 +1,890 @@
+  * An actor's peer simulator can be as simple as a pre-registered stream of messages,
+    or as sophisticated as an error injector.
+  * In a pure actor model setting, all interactions with the system are done via actors,
+    and can be simulated, including the clock and the file system.
+
+Thin-edge leverages Rust, adding robustness to the actor model.
+
+* __Robustness__
+  * No message can be sent if not understood by the recipient.
+    The compatibility, between the messages sent to an actor
+    and the messages this actor can actually process,
+    is verified at compile-time.
+
+However, using Rust also introduce specific challenges, as we will see below.
+To name just a few:
+
+* As all the message types have to be known at compile time,
+  we need to be cautious not to make all the actor *implementations* depending on each others,
+  due to dependencies at the *message* level.
+  * The implementation of an actor that consumes telemetry data
+    must not depend on the implementations of actors producing telemetry data - and vice versa.
+* As no message can be sent without being understood by the recipient,
+  we need a shared definition of IoT messages.
+  * Consumers and producers of telemetry data must agree on what is telemetry data.
+* However, we also expect the system to be extended by independent vendors,
+  and thin-edge should not pre-defined all the messages that can be exchanged by actors.
+  * A contributor should be able to define its own set of messages usable by others.  
+
+## From actors to software components
+
+Using actors is key but not enough.
+Yes, the actor model helps thin-edge rust developers to implement components in isolation.
+However, thin-edge also aims to provide flexibility to the agent developers with ready-to-use executables,
+that can be tuned on-site for a specific use-case,
+by activating only specific features among all those available.
+
+The design must address not only the API for *actors* but also the packaging of these actors into *extensions*,
+their integration into *executables* and their *configuration* by agent developers and end-users. 
+
+* Elementary features are provided by __actors__: units of computation that interact with asynchronous messages.
+* Actors are packaged into __extensions__: rust libraries that provide implementations along interfaces,
+  input and output messages as well as instantiation services.
+* A thin-edge __executable__ makes available a collection of extensions ready to be enabled on-demand.
+* A user provided __configuration__ enables and configures sherry-picked extensions making an agent.
+
+To be discussed:
+* Do we really need to design *now* how the users could sherry-pick extensions from a batteries-included software?
+  This could be done in a second step, starting with static assemblages of actors.
+
+## Requirements
+
+One should be able to build a thin-edge executable from extensions that have been implemented independently.
+
```

Comment:
Maybe an additional requirement (if it is not covered by an existing point)

```suggestion
* Actors should be composable and support being synchronous. This would allow the reusing of existing actors in other actors to build an overarching orchestrator.  For example, Actor A receives a message, sends a message to Actor B to let the existing actor do some work, waits for Actor B to finish, then Actors A sends a message that it is finished
```

A concrete example of this is the handling of Cumulocity Device Profile operations. This operation contains the same contents as the software, configuration and firmware operations, but it rolled into one operation, where each type of item is processed serially (to avoid conflicting side-effects). It would make sense if the device profile actor could reuse the same logic already implemented for the software, configuration and firmware handling.

Reply:
This is an important point ... but for which I struggled to find a solution.

This is required for two reasons:
* to avoid conflicting concurrent operations, as you said
* but also, to have simpler code. To manage concurrent operations, an actor has to maintain some specific state for each on-going operation and to store and resume these states on each context switch. Hence having the ability for an actor to pause awaiting the response of another is key to mix event-loop and async/await programming style.  

=> I will add this requirement to the doc.

Reply:
Added.